### DISCUSSION TITLE
Why should `minTTL` be equals to `( 3 * ElectionTicks) / 2 * heartbeat`

### DISCUSSION THREAD

**Discussion by Phoenix500526** [etcd-io/etcd#15858] [Q&A] [answered]
Hi, everyone. I'm a newbie to etcd and have a problem with some code implemented in etcd. 

Let's look at some codes in `NewServer` method (reference [NewServer](URL[)](https://github.com/etcd-io/etcd/blob/f6161673af46ea925ed016d5f2cb69e787241069/server/etcdserver/server.go#L306) )
```go
// NewServer creates a new EtcdServer from the supplied configuration. The
// configuration is considered static for the lifetime of the EtcdServer.
func NewServer(cfg config.ServerConfig) (srv *EtcdServer, err error) {
	// omit some irrelevant codes....
	heartbeat := time.Duration(cfg.TickMs) * time.Millisecond
	srv = &EtcdServer{
		readych:               make(chan struct{}),
		Cfg:                   cfg,
		lgMu:                  new(sync.RWMutex),
		lg:                    cfg.Logger,
		errorc:                make(chan error, 1),
		v2store:               b.storage.st,
		snapshotter:           b.ss,
		r:                     *b.raft.newRaftNode(b.ss, b.storage.wal.w, b.cluster.cl),
		memberId:              b.cluster.nodeID,
		attributes:            membership.Attributes{Name: cfg.Name, ClientURLs: cfg.ClientURLs.StringSlice()},
		cluster:               b.cluster.cl,
		stats:                 sstats,
		lstats:                lstats,
		SyncTicker:            time.NewTicker(500 * time.Millisecond),
		peerRt:                b.prt,
		reqIDGen:              idutil.NewGenerator(uint16(b.cluster.nodeID), time.Now()),
		AccessController:      &AccessController{CORS: cfg.CORS, HostWhitelist: cfg.HostWhitelist},
		consistIndex:          b.storage.backend.ci,
		firstCommitInTerm:     notify.NewNotifier(),
		clusterVersionChanged: notify.NewNotifier(),
	}
	serverID.With(prometheus.Labels{"server_id": b.cluster.nodeID.String()}).Set(1)
	srv.cluster.SetVersionChangedNotifier(srv.clusterVersionChanged)
	srv.applyV2 = NewApplierV2(cfg.Logger, srv.v2store, srv.cluster)

	srv.be = b.storage.backend.be
	srv.beHooks = b.storage.backend.beHooks
	minTTL := time.Duration((3*cfg.ElectionTicks)/2) * heartbeat

	// always recover lessor before kv. When we recover the mvcc.KV it will reattach keys to its leases.
	// If we recover mvcc.KV first, it will attach the keys to the wrong lessor before it recovers.
	srv.lessor = lease.NewLessor(srv.Logger(), srv.be, srv.cluster, lease.LessorConfig{
		MinLeaseTTL:                int64(math.Ceil(minTTL.Seconds())),
		CheckpointInterval:         cfg.LeaseCheckpointInterval,
		CheckpointPersist:          cfg.LeaseCheckpointPersist,
		ExpiredLeasesRetryInterval: srv.Cfg.ReqTimeout(),
	})

        // omit some irrelevant codes....
}
```

And my question is, as the title describes, why should the `minTTL` be equals to `( 3 * ElectionTicks) / 2 * heartbeat`. 

Can anybody tell me why? Thanks in advance!

**Comment by Phoenix500526** [comment]
@ahrtr @tjungblu @serathius @jmhbnz

**Comment by ahrtr** [comment, marked as answer]
The main purpose is to ensure lease will not expire on leader failover. FYI. https://github.com/etcd-io/etcd/issues/6025#issuecomment-236961028. 

The [randomElectionTimeout](https://github.com/etcd-io/raft/blob/177ef28aae851bb644181bf531793ee33a895c56/raft.go#L1901) is in `[electionTimeout, 2*ElectionTimeout)`, so I guess it's the reason why the minTTL was set to a value in the middle (`1.5*ElectionTimeout`) in https://github.com/etcd-io/etcd/pull/6085.

**Reply by ahrtr** [reply]
I think it might make more sense to set minTTL to the max randonElectionTimeout, namely 2*electionTimeout. @mitake WDYT?

**Reply by mitake** [reply]
@ahrtr I guess it's worth examining. But it's possible to have failed elections, so utilizing election timeout for determining min TTL of lease is not enough for avoiding lease expiration during failover. I think it's better to make every state transition managed by Raft messages (like what you are working on in https://github.com/etcd-io/etcd/pull/14094) is better.

**Reply by ahrtr** [reply]
> But it's possible to have failed elections, so utilizing election timeout for determining min TTL of lease is not enough for avoiding lease expiration during failover. I think it's better to make every state transition managed by Raft messages (like what you are working on in https://github.com/etcd-io/etcd/pull/14094) is better.

Yes, it's a big topic which can be discussed separately. Thanks @mitake .

**Reply by Phoenix500526** [reply]
Thank you for your answers 😀
